### PR TITLE
Runtime-rs: support persist file

### PR DESCRIPTION
**Pull request by Tim-0731-Hzt**
This two commits has been approved and merged into the runtime-rs branch this week, so this pr is to merge that into the main. It supports the functionality of save, restore and cleanup.

Fixes:#4891
Signed-off-by: Zhongtao Hu <zhongtaohu.tim@linux.alibaba.com>

### CODE REVIEW

**Review by quanweiZhou — APPROVED** [review]
Thanks @Tim-0731-Hzt， LGTM

**Review by liubin — APPROVED** [review]

File: src/runtime-rs/crates/resource/src/manager_inner.rs [hunk 1]
```diff
@@ -198,3 +202,48 @@ impl ResourceManagerInner {
...
+            config: resource_args.config,
+        };
+        Ok(Self {
+            sid: resource_args.sid,
+            agent: resource_args.agent,
+            hypervisor: resource_args.hypervisor,
+            network: None,
+            share_fs: None,
+            rootfs_resource: RootFsResource::new(),
+            volume_resource: VolumeResource::new(),
```
Comment (jodh-intel) [inline src/runtime-rs/crates/resource/src/manager_inner.rs:240]: It's a shame that not all the new types implement Default as the code could be simplified if they did.

File: src/runtime-rs/crates/runtimes/src/manager.rs [hunk 2]
```diff
@@ -127,8 +130,36 @@ impl RuntimeHandlerManager {
...
+        let sandbox_args = SandboxRestoreArgs {
+            sid: inner.id.clone(),
+            toml_config: TomlConfig::default(),
+            sender,
+        };
+        match sandbox_state.sandbox_type {
+            SandboxTYPE::VIRTCONTAINER => {
+                let sandbox = VirtSandbox::restore(sandbox_args, sandbox_state)
+                    .await
+                    .context("failed to retore the sandbox")?;
```
Comment (jodh-intel) [inline src/runtime-rs/crates/runtimes/src/manager.rs:147]: Typo:

```suggestion
                    .context("failed to restore the sandbox")?;
```

File: src/runtime-rs/crates/runtimes/src/manager.rs [hunk 3]
```diff
@@ -127,8 +130,36 @@ impl RuntimeHandlerManager {
...
+                sandbox
+                    .cleanup(&inner.id)
+                    .await
+                    .context("failed to cleanup the resource")?;
+            }
+            SandboxTYPE::LINUXCONTAINER => {
+                // TODO :support linux container
```
Comment (jodh-intel) [inline src/runtime-rs/crates/runtimes/src/manager.rs:154]: Nit: Ideally, you'd specify a GH issue URL here.

File: src/runtime-rs/crates/runtimes/src/manager.rs [hunk 4]
```diff
@@ -127,8 +130,36 @@ impl RuntimeHandlerManager {
...
+                return Ok(());
+            }
+            SandboxTYPE::WASMCONTAINER => {
+                // TODO :support wasm container
```
Comment (jodh-intel) [inline src/runtime-rs/crates/runtimes/src/manager.rs:158]: Nit: Ideally, you'd specify a GH issue URL here.

**Review by jodh-intel — APPROVED** [review]
Thanks @Tim-0731-Hzt - just a few nits, but overall...

lgtm

File: src/runtime-rs/crates/resource/src/manager_inner.rs (hunk 1 above)
Comment (Tim-0731-Hzt) [reply src/runtime-rs/crates/resource/src/manager_inner.rs:240]: As the functionality of cleanup is not fully support at this stage, this is pr is aiming to give a framework for that. The problem you mentioned would be solved in the future.

File: src/runtime-rs/crates/runtimes/src/manager.rs (hunk 2 above)
Comment (Tim-0731-Hzt) [reply src/runtime-rs/crates/runtimes/src/manager.rs:147]: Thanks, fixed

File: src/runtime-rs/crates/runtimes/src/manager.rs (hunk 3 above)
Comment (Tim-0731-Hzt) [reply src/runtime-rs/crates/runtimes/src/manager.rs:154]: fixed

File: src/runtime-rs/crates/runtimes/src/manager.rs (hunk 4 above)
Comment (Tim-0731-Hzt) [reply src/runtime-rs/crates/runtimes/src/manager.rs:158]: fixed

File: src/runtime-rs/crates/hypervisor/src/dragonball/inner.rs [hunk 5]
```diff
@@ -307,3 +310,43 @@ impl DragonballInner {
         self.config.clone()
     }
 }
+
+#[async_trait]
+impl Persist for DragonballInner {
+    type State = HypervisorState;
+    type ConstructorArgs = ();
+    /// Save a state of the component.
+    async fn save(&self) -> Result<Self::State> {
```
[the remark this answers is not in the document]
Comment (Tim-0731-Hzt) [reply src/runtime-rs/crates/hypervisor/src/dragonball/inner.rs:319]: fixed

**Pull request merged**